File: MAT DISCRETA/CRUZAMENTO.cpp
#include <stdio.h>
#include <stdlib.h>
#include <String.h>

#include <stdio.h>
#include <stdlib.h>

void cruzamento() {
    // Tempo de cada fase
    int paradoP = 40, esperaP = 3, andandoP = 55;
    int paradoS = 47, esperaS = 3, andandoS = 39;

    while (1) {
        // *** Via Principal Andando - Via Secundária Parada ***
        for (int i = paradoS; i > 0; i--) {
            printf("Via Principal       Via Secundaria\n");
            printf("Andando: %d         Parado: %d\n", andandoP--, i);
            if (andandoP == 0) break;
        }

        // Tempo de espera sincronizado antes de inverter o fluxo
        for (int j = esperaP; j > 0; j--) {
            printf("Via Principal       Via Secundaria\n");
            printf("Espera: %d          Espera: %d\n", j, j);
        }

        // *** Via Secundária Andando - Via Principal Parada ***
        for (int i = paradoP; i > 0; i--) {
            printf("Via Principal       Via Secundaria\n");
            printf("Parado: %d          Andando: %d\n", i, andandoS--);
            if (andandoS == 0) break;
        }

        // Tempo de espera sincronizado antes de inverter o fluxo novamente
        for (int j = esperaS; j > 0; j--) {
            printf("Via Principal       Via Secundaria\n");
            printf("Espera: %d          Espera: %d\n", j, j);
        }

        // Reinicia os tempos para o próximo ciclo
        andandoP = 55;
        andandoS = 39;
    }
}

int main() {
    cruzamento();
    return 0;
}


/*int viaP(){
	int parado, espera, andando; 
	parado = 40;
	espera = 3;
	andando = 55;
	
	for (int i = 1; i <= parado; i++){
		printf("Tempo parado: %d\n", i);
		
		if(i ==  40){
			for(int j = 1; j <= espera; j++){
				printf("Tempo de espera: %d\n", j);
			}
			
			if(i ==  3){
				for(int k = 1; k <= andando; k++){
					printf("Tempo de andando: %d\n", k);
				}		
			}
		}
	}
	return 0; 
}


int viaS(){
	int parado, espera, andando; 	
	parado = 47;
	espera = 3;
	andando = 39;
	
	for (int i = 1; i <= parado; i++){
		printf("Tempo parado: %d\n", i);
		
		if(i ==  47){
			for(int j = 1; j <= espera; j++){
				printf("Tempo de espera: %d\n", j);
			}
			
			if(i ==  3){
				for(int k = 1; k <= andando; k++){
					printf("Tempo andando: %d\n", k);
				}		
			}
		}
	}
	return 0;
}
	//printf("Via Principal: %i      Via Secundaria: %i\n\n", viaP(), viaS());

int main(){
	
	for(int i = 0; i < 5; i++){
			printf("Via Principal             Via Secundaria\n\n");
			printf("%i                      %i\n",viaP(),viaS());
	}

	return 0;
}*/

File: MAT DISCRETA/CRUZAMENTO/src/Cruzamento.java
public class Cruzamento {

    public static void main(String[] args) {
        cruzamento();
    }

    public static void cruzamento() {
        // Limites e tempos para as vias
        int limiteParadoP = 60;  // Limite máximo de parada para a via principal
        int paradoP = 40, esperaP = 3, andandoP = 55;
        int paradoS = 47, esperaS = 3, andandoS = 39;

        while (true) {
            // *** Via Principal Andando - Via Secundária Parada ***
            for (int i = paradoS; i > 0; i--) {
                System.out.println("Via Principal       Via Secundária");
                System.out.printf("Andando: %d         Parado: %d\n", andandoP--, i);
                if (andandoP == 0) break;  // Sai do loop se a via principal terminar de andar
                dormir(1000); // Pausa de 1 segundo para simular o tempo real
            }

            // Tempo de espera sincronizado
            for (int j = esperaP; j > 0; j--) {
                System.out.println("Via Principal       Via Secundária");
                System.out.printf("Espera: %d          Espera: %d\n", j, j);
                dormir(1000);
            }

            // *** Via Secundária Andando - Via Principal Parada ***
            int tempoParadoP = 0;
            for (int i = paradoP; i > 0; i--) {
                System.out.println("Via Principal       Via Secundária");
                System.out.printf("Parado: %d          Andando: %d\n", i, andandoS--);

                tempoParadoP++;
                if (andandoS == 0 || tempoParadoP >= limiteParadoP) break;  
                // Sai se a via secundária terminar de andar ou atingir o limite
                dormir(1000);
            }

            // Tempo de espera sincronizado
            for (int j = esperaS; j > 0; j--) {
                System.out.println("Via Principal       Via Secundária");
                System.out.printf("Espera: %d          Espera: %d\n", j, j);
                dormir(1000);
            }

            // Reinicia os tempos para o próximo ciclo
            andandoP = 55;
            andandoS = 39;
        }
    }

    // Método auxiliar para simular o tempo de espera (em milissegundos)
    public static void dormir(int tempo) {
        try {
            Thread.sleep(tempo);
        } catch (InterruptedException e) {
            Thread.currentThread().interrupt();
        }
    }
}
